calico_rkt/calico_policy.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2015 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from constants import *
+from calico_rkt import configure_logging
+from pycalico.datastore import DatastoreClient
+from pycalico.datastore_datatypes import Rule, Rules
+
+KUBERNETES_DEFAULT_PROFILE = "default_profile"
+
+_log = logging.getLogger(__name__)
+configure_logging(_log)
+
+class BasePolicyDriver(object):
+    """
+    Abstract base class of a CNI Policy Driver.
+
+    The CNI Policy Driver is responsible for applying network policy to
+    networked containers and creating/applying Calico profiles to Calico
+    endpoints.
+    """
+    def __init__(self):
+
+        self._client = DatastoreClient()
+        """
+        DatastoreClient for access to the Calico datastore.
+        """
+
+        self.profile_name = None
+        """
+        Name of profile for attach to endpoint. Must be set in init function
+        of subclass.
+        """
+
+    def set_profile(self, endpoint):
+        """Sets a profile for the networked container on the given endpoint.
+
+        Create a profile if it is not yet created.
+
+        :param endpoint:
+        :return:
+        """
+        if not self.profile_name:
+            raise NotImplementedError("A profile name has not been set.")
+        if not self._client.profile_exists(self.profile_name):
+            # If the profile doesn't exist, create it.
+            _log.info("Creating new profile '%s'", self.profile_name)
+            rules = self.generate_rules()
+            self._client.create_profile(self.profile_name, rules)
+
+        # Set the default profile on this pod's Calico endpoint.
+        _log.info("Setting profile '%s' on endpoint %s",
+                  self.profile_name, endpoint.endpoint_id)
+        self._client.set_profiles_on_endpoint(profile_names=[self.profile_name],
+                                              endpoint_id=endpoint.endpoint_id)
+
+    def remove_profile(self):
+        """Right now this function is no-op.
+
+        Need to think more about how to handle the race condition that exists
+        between removing a profile and creating a new one of the same name.
+
+        For now we'll leak profiles in Calico datastore.
+
+        :return:
+        """
+        _log.info("Called remove profile. Doing nothing for now.")
```

Comment:
This should still raise a NotImplementedError - sub-classes need to explicitly override this.